fuel-core/src/executor.rs

```diff
@@ -603,29 +598,38 @@ impl Executor {
             }
         }
 
-        let tx = vm_result.tx().clone().into();
-        // Store tx into the block db transaction
-        tx_db_transaction
-            .deref_mut()
-            .storage::<Transactions>()
-            .insert(&tx_id, &tx)?;
-
         // change the spent status of the tx inputs
-        self.spend_inputs(
-            vm_result.tx(),
+        self.spend_input_utxos(
+            &tx,
             tx_db_transaction.deref_mut(),
             *header.height(),
             self.config.utxo_validation,
         )?;
 
         // persist any outputs
-        self.persist_outputs(
+        self.persist_output_utxos(
             *header.height(),
             &tx_id,
             tx_db_transaction.deref_mut(),
             vm_result.tx().inputs(),
             vm_result.tx().outputs(),
         )?;
+        self.compute_outputs(
```

Comment:
It looks like we're persisting the outputs before we compute their state/balance roots etc. If we had tests to check that the balance / state roots on contract outputs are set and persisted correctly, this would've been caught.

Reply:
We have a test=) `persist_output_utxos` persist only outputs related to utxo=)